src/io/future.hpp

```diff
@@ -0,0 +1,257 @@
+          // the simulator of our waiting_ status.
+          break;
+        }
+      }
+      if (!simulator_progressed) {
+        cv_.wait(lock);
+      }
+      MG_ASSERT(!consumed_, "Future consumed twice!");
+    }
+
+    T ret = std::move(item_).value();
+    item_.reset();
+
+    waiting_ = false;
+    consumed_ = true;
+
+    return ret;
+  }
+
+  bool IsReady() {
+    std::unique_lock<std::mutex> lock(mu_);
+    return item_;
+  }
+
+  std::optional<T> TryGet() {
+    std::unique_lock<std::mutex> lock(mu_);
+
+    if (item_) {
+      T ret = std::move(item_).value();
+      item_.reset();
+
+      waiting_ = false;
+      consumed_ = true;
+
+      return ret;
+    } else {
+      return std::nullopt;
+    }
+  }
+
+  void Fill(T item) {
+    {
+      std::unique_lock<std::mutex> lock(mu_);
+
+      MG_ASSERT(!consumed_, "Promise filled after it was already consumed!");
+      MG_ASSERT(!item_, "Promise filled twice!");
+
+      item_ = item;
+    }  // lock released before condition variable notification
+
+    cv_.notify_all();
+  }
+
+  bool IsAwaited() {
+    std::unique_lock<std::mutex> lock(mu_);
+    return waiting_;
+  }
+};
+}  // namespace
+
+template <typename T>
+class Future {
+  bool consumed_or_moved_ = false;
+  std::shared_ptr<Shared<T>> shared_;
+
+ public:
+  explicit Future(std::shared_ptr<Shared<T>> shared) : shared_(shared) {}
+
+  Future() = delete;
+  Future(Future &&old) {
+    shared_ = std::move(old.shared_);
+    consumed_or_moved_ = old.consumed_or_moved_;
+    MG_ASSERT(!old.consumed_or_moved_, "Future moved from after already being moved from or consumed.");
+    old.consumed_or_moved_ = true;
+  }
+  Future &operator=(Future &&old) {
+    shared_ = std::move(old.shared_);
+    MG_ASSERT(!old.consumed_or_moved_, "Future moved from after already being moved from or consumed.");
+    old.consumed_or_moved_ = true;
+  }
+  Future(const Future &) = delete;
+  Future &operator=(const Future &) = delete;
+  ~Future() = default;
+
+  /// Returns true if the Future is ready to
+  /// be consumed using TryGet or Wait (prefer Wait
+  /// if you know it's ready, because it doesn't
+  /// return an optional.
+  bool IsReady() {
+    MG_ASSERT(!consumed_or_moved_, "Called IsReady after Future already consumed!");
+    return shared_->IsReady();
+  }
+
+  /// Non-blocking method that returns the inner
+  /// item if it's already ready, or std::nullopt
+  /// if it is not ready yet.
+  std::optional<T> TryGet() {
+    MG_ASSERT(!consumed_or_moved_, "Called TryGet after Future already consumed!");
+    std::optional<T> ret = shared_->TryGet();
+    if (ret) {
+      consumed_or_moved_ = true;
+    }
+    return ret;
+  }
+
+  /// Block on the corresponding promise to be filled,
+  /// returning the inner item when ready.
+  T Wait() {
+    MG_ASSERT(!consumed_or_moved_, "Future should only be consumed with Wait once!");
+    T ret = shared_->Wait();
+    consumed_or_moved_ = true;
+    return ret;
+  }
```

Comment:
in general I like the idea of enforcing move-only semantics around Future and Promise as much as possible. Is there a reason not to simply make TryGet also an rvalue reference function?

Reply:
My only reasoning for this `TryGet` _might_ consume the object, but not certainly. Because of that we might and up finding with the compiler, because it cannot figure out in compilation time whether the `Future` will be moved or not, thus popping up a lot of "use after move" warnings.